source-mongodb/pull.go

```diff
@@ -0,0 +1,267 @@
+	if err != nil {
+		return fmt.Errorf("connecting to database: %w", err)
+	}
+	log.Info("connected to database")
+
+	defer func() {
+		if err = client.Disconnect(ctx); err != nil {
+			panic(err)
+		}
+	}()
+
+	var c = capture{
+		Output: &boilerplate.PullOutput{Stream: stream},
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	if err := c.Output.Ready(); err != nil {
+		return err
+	}
+
+	for idx, binding := range open.Open.Capture.Bindings {
+		var res resource
+		if err := pf.UnmarshalStrict(binding.ResourceSpecJson, &res); err != nil {
+			return fmt.Errorf("parsing resource config: %w", err)
+		}
+		var resState, _ = prevState.Resources[resourceId(res)]
+
+		// Only backfill the first time, and we know that when there is no state
+		if len(prevState.Resources) == 0 {
+			if err := c.BackfillCollection(ctx, client, uint32(idx), res); err != nil {
+				return fmt.Errorf("backfill: %w", err)
+			}
+		}
+
+		eg.Go(func() error { return c.ChangeStream(ctx, client, uint32(idx), res, resState) })
+	}
+
+	if err := eg.Wait(); err != nil && !errors.Is(err, io.EOF) {
+		return err
+	}
+
+	return nil
+}
+
+type capture struct{
+	Output   *boilerplate.PullOutput
+	BackfillFinishedAt *primitive.Timestamp
+}
+
+type captureState struct{
+	Resources map[string]bson.Raw `json:"resources"`
+}
+
+func (s *captureState) Validate() error {
+	return nil
+}
+
+type changeEvent struct{
+	OperationType string `bson:"operationType"`
+	FullDocument bson.M `bson:"fullDocument"`
+}
+
+const changeStreamFatalErrorCode = 280
+const resumePointGoneErrorMessage = "the resume point may no longer be in the oplog"
+
+func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, binding uint32, res resource, resumeToken bson.Raw) error {
+	var db = client.Database(res.Database)
+
+	var collection = db.Collection(res.Collection)
+
+	log.Debug("listening on changes on collection ", res.Collection)
+	var eventFilter = bson.D{{"$match", bson.D{{"$or",
+	bson.A{
+		// TODO: handle deletion events
+		// bson.D{{"operationType", "delete"}}
+		bson.D{{"operationType", "insert"}},
+		bson.D{{"operationType", "update"}},
+	}}}}}
+	var opts = options.ChangeStream().SetFullDocument(options.UpdateLookup)
+	if resumeToken != nil {
+		opts = opts.SetResumeAfter(resumeToken)
+	} else if c.BackfillFinishedAt != nil {
+		opts = opts.SetStartAtOperationTime(c.BackfillFinishedAt)
+	}
+	cursor, err := collection.Watch(ctx, mongo.Pipeline{eventFilter}, opts)
+	if err != nil {
+		// This error means events from the starting point are no longer available,
+		// which means we have hit a gap in between last run of the connector and
+		// this one. We re-set the checkpoint and error out so the next run of the
+		// connector will run a backfill.
+		if e, ok := err.(mongo.ServerError); ok {
+			if e.HasErrorCode(changeStreamFatalErrorCode) && e.HasErrorMessage(resumePointGoneErrorMessage) {
+				if err = c.Output.Checkpoint([]byte("{}"), false); err != nil {
```

Comment:
This will erase checkpoint state for every collection, right?

I guess that in general all change streams of a capture will be resuming from the same point in time in the oplog, but I'm wondering if for instance it's possible for a collection which changes really infrequently to have an older cursor value (since it looks like we only get a new cursor value alongside a document change event), such that it can't be resumed any longer after a restart but we'd probably prefer for other more recent (which I think would tend to be higher-volume and more massive) collections of the capture to not have to redo the backfill?